lib/ansible/module_utils/api.py

```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
     return wrapper
 
 
-def retry(retries=None, retry_pause=1):
-    """Retry decorator"""
-    def wrapper(f):
+def _retry_limit_exceeded(retries):
+    def function_wrapper(function):
+        def raise_error(*args, **kwargs):
+            raise Exception("Retry limit exceeded: %d" % retries)
+        return raise_error
+    return function_wrapper
 
-        def retried(*args, **kwargs):
-            retry_count = 0
-            if retries is not None:
-                ret = None
-                while True:
-                    retry_count += 1
-                    if retry_count >= retries:
-                        raise Exception("Retry limit exceeded: %d" % retries)
-                    try:
-                        ret = f(*args, **kwargs)
-                    except Exception:
-                        pass
-                    if ret:
-                        break
-                    time.sleep(retry_pause)
-                return ret
-
-        return retried
-    return wrapper
+
+def retry(retries=None, retry_pause=1):
+    """Retry decorator.
+    This assumes the function has already been called, so if retries == 1, the maximum attempts
+    have been reached. If retries is None, the function is a no-op.
+    """
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
+    if retries <= 1:
+        return _retry_limit_exceeded(retries)
```

Comment:
This is a bit messy. You can't have 'raise Exception in a lambda, so I made another function just to do that. I'm of half a mind to leave `retry` as it was rather than make it use the common function since it was doing some not-so-generic things.

Reply:
Removed this from the PR.